jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -2579,6 +2581,215 @@ ecma_builtin_array_prototype_object_copy_within (const ecma_value_t args[], /**<
 
   return ecma_copy_value (ecma_make_object_value (obj_p));
 } /* ecma_builtin_array_prototype_object_copy_within */
+
+/**
+ * FlattenIntoArray
+ *
+ * See also:
+ *          ECMA-262 v10, 22.1.3.10.1
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_flatten_into_array (ecma_value_t target,
+                                       ecma_object_t *source,
+                                       uint32_t sourceLen,
+                                       uint32_t start,
+                                       ecma_number_t depth,
+                                       ecma_value_t mapped_value,
+                                       ecma_value_t thisArg)
+{
+  uint32_t targetIndex = start;
+  uint32_t sourceIndex = 0;
+  bool is_mapped_value_not_undefined = !ecma_is_value_undefined (mapped_value);
+
+  while (sourceIndex < sourceLen)
+  {
+    ecma_value_t element = ecma_op_object_find_by_uint32_index (source, sourceIndex);
+    if (ECMA_IS_VALUE_ERROR (element))
+    {
+      return element;
+    }
+
+    if (ecma_is_value_found (element))
+    {
+      if (is_mapped_value_not_undefined)
+      {
+        ecma_value_t sourceVal = ecma_make_uint32_value (sourceIndex);
+        ecma_value_t args[] = {element,  sourceVal, ecma_make_object_value (source)};
+        ecma_value_t temp_element = ecma_op_function_call (ecma_get_object_from_value (mapped_value), thisArg, args, 3);
+
+        ecma_free_value (element);
+        ecma_free_value (sourceVal);
+
+        if (ECMA_IS_VALUE_ERROR (temp_element))
+        {
+          return temp_element;
+        }
+
+        element = temp_element;
+      }
+      bool shouldFlatten = false;
+      if (depth > 0)
+      {
+        ecma_value_t is_array = ecma_is_value_array (element);
+
+        if (ECMA_IS_VALUE_ERROR (is_array))
+        {
+          ecma_free_value (element);
+          return is_array;
+        }
+        shouldFlatten = ecma_is_value_true (is_array);
+      }
+
+      if (shouldFlatten)
```

Comment:
This if block could be moved into the previous if block, then there's no need for the bool variable.